core/node/libp2p/routing.go

```diff
@@ -127,6 +130,56 @@ func BaseRouting(experimentalDHTClient bool) interface{} {
 	}
 }
 
+type delegatedRouterOut struct {
```

Comment:
Unfortunately, this isn't all the fx wrestling to do for this to be "right". We currently have two main routers we make use of:
1. DHT (internally is WAN + LAN, but whatever)
2. PubSub (used as a router for IPNS records)

PubSub uses the DHT for topic discovery through via the `BaseIPFSRouter` this ideally should change to support Reframe as well. https://github.com/ipfs/go-ipfs/blob/ef866a1400b3b2861e5e8b6cc9edc8633b890a0a/core/node/libp2p/topicdiscovery.go#L16

Note: You can't just set the BaseIPFSRouter to be the combined Reframe + DHT routers because the Batched provider (used by the accelerated DHT client for bulk publishing) does a Go interface check on the router. https://github.com/ipfs/go-ipfs/blob/e8f1ce07b3d8076da5baedd772d80368804e57a7/core/node/provider.go#L73

Reply:
Thanks for the heads up. Some questions about this:
PubSub is a partial Routing implementation (only implementing ContentRouting) that needs a Routing implementation to work, DHT in this case (BaseIpfsRouting === Routing), right? Do we want to be able to use our Delegated Routers on any of these Routing implementations? Is that why you suggested a `Methods` param on the configuration?

Reply:
> Is that why you suggested a Methods param on the configuration?

`Methods` isn't strictly required and I'm fine dropping it until a user request comes in. The two use cases I was thinking it might cover are:
1. While router X can do A, B, C I only want to use it for A and B. Perhaps limiting routers in this way won't be a common thing and we can just wait for users to ask for it.
2. While Reframe supports A, B, C this router only supports B and I don't want to rely on the `Identify` method or caching errors from having tried A and C and them failing. As indicated this isn't strictly required because we have other mechanisms available.

Reply:
> PubSub is a partial Routing implementation (only implementing ContentRouting) that needs a Routing implementation to work, DHT in this case (BaseIpfsRouting === Routing), right? Do we want to be able to use our Delegated Routers on any of these Routing implementations?

PubSub is really just used to satisfy the ValueStore interface (particularly for Put/Get for IPNS records) but it requires somehow connecting to other peers interested in those pubsub topics. While this could happen by accident it's much more reliable to use another routing system (e.g. one like the DHT that supports provider records) to discover other peers interested in those topics.

I think it makes sense to leverage any of the routing systems that support discovery (e.g. provider records) to perform discovery for the pubsub topics. This means any of the delegated routers should be usable here.

A guiding thought here might be that ideally a user should be able to turn off the DHT entirely within go-ipfs and outsource those requests to some other routing system (it might even be a DHT daemon running on localhost). At the moment we can't replicate the full behavior of the DHT via Reframe for a couple of reasons, but this should illustrate the point.

----

Reasons why we can't replicate the full behavior of the DHT via reframe:
- There are no methods for advertising provider records, and even if there were the IPFS Public DHT cannot currently support delegated advertising of provider records at all. https://github.com/libp2p/go-libp2p-kad-dht/issues/584
- There are no methods for advertising peer records, and even if there were the IPFS Public DHT cannot currently support delegated advertising of peer records at all. See issue for provider records
- There are no methods for retrieving peer records. The IPFS Public DHT can support this we just need to land the Reframe changes https://github.com/ipfs/specs/pull/279

Reply:
Added more fx magic to create a TieredRouting from all Routers on `routers` group.